### PR TITLE
Forward port backend improvements

### PR DESCRIPTION
**Pull request by sjrd**
Forward port of

* #10022
* #10025

### CODE REVIEW

File: test/junit/scala/tools/nsc/backend/jvm/analysis/TypeFlowAnalyzerTest.scala [hunk 1]
```diff
@@ -50,13 +50,14 @@ class TypeFlowAnalyzerTest extends BytecodeTesting {
...
     val a = new NonLubbingTypeFlowAnalyzer(m, "C")
-    val v = a.frameAt(findInstr(m, "ARETURN")).getStack(0)
-    assertEquals(v.getType, BasicValue.REFERENCE_VALUE.getType)
-    assert(!v.isInstanceOf[AaloadValue])
-    assertEquals(a.preciseAaloadTypeDesc(v), "Ljava/lang/Object;")
+    for (instr <- findInstrs(m, "ARETURN")) {
+      val v = a.frameAt(instr).getStack(0)
+      assertEquals(v.getType, BasicValue.REFERENCE_VALUE.getType)
+      assert(v.isInstanceOf[AaloadValue])
+      assertEquals(a.preciseAaloadTypeDesc(v), "Ljava/lang/String;")
```
Comment (sjrd) [inline test/junit/scala/tools/nsc/backend/jvm/analysis/TypeFlowAnalyzerTest.scala:59]: This change from `java.lang.Object` to `java.lang.String` is a progression on this particular example. However, I suspect that this change may void the original intent of this test case. @lrytz Could you take a particular look here?

**Comment by sjrd** [comment]
Ping @lrytz ?

**Comment by lrytz** [comment]
I think @SethTisue wanted to wait with this one until after 2.13.9.

**Comment by sjrd** [comment]
Oh, did he? (did you?) I see he specifically assigned this PR to 2.13.9. There was https://github.com/scala/scala/pull/10028 which he definitely wanted to delay to 2.12.17 (and which I should rebase, now that 2.12.16 is out, actually).

**Comment by lrytz** [comment]
Ah I confused the two, sorry. Will look into it.

**Review by lrytz — APPROVED** [review]

**Comment by lrytz** [comment]
Thanks @sjrd!

**Pull request merged**